src/Symfony/Component/HttpClient/ConditionalHttpClient.php

```diff
@@ -0,0 +1,79 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient;
+
+use phpDocumentor\Reflection\Types\Boolean;
+use Symfony\Component\HttpClient\Exception\InvalidArgumentException;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+use Symfony\Contracts\HttpClient\ResponseStreamInterface;
+
+/**
+ * Auto-configure the default options based on the requested URL.
+ *
+ * @author Anthony Martin <anthony.martin@sensiolabs.com>
+ *
+ * @experimental in 4.3
+ */
+class ConditionalHttpClient implements HttpClientInterface
+{
+    use HttpClientTrait;
+
+    private $client;
+    private $options;
+    private $absoluteUrls;
+
+    /**
+     * @param array[] $options the default options to use when the regexp provided as key matches the requested URL
+     * @param bool    $absoluteUrls indicates if at least one rtegexp url given in $options would be on absolute url, in this case we need to check the url give to request() with use of prepareRequest to parse tu url etc...
+     */
+    public function __construct(HttpClientInterface $client, array $options, bool $absoluteUrls = false)
+    {
+        $this->client = $client;
+        $this->options = $options;
+        $this->absoluteUrls = $absoluteUrls;
```

Comment:
I think we should remove this and mandate using absolute URLs in `request`:
```php
if (!parse_url($url, PHP_URL_SCHEME) || !parse_url($url, PHP_URL_HOST)) {
    throw new InvalidArgumentException(sprintf('Unsupported URL: host or scheme is missing in "%s".', $url));
}
```
This will make things much simpler. Then, another HttpClient decorator could easilly allow turning relative to absolute on top of this one:

NativeHttpClient -> ConditionalHttpClient -> BaseUriHttpClient (random name - for another PR)